modules/registrars/openprovider/Controllers/System/DnsController.php

```diff
@@ -45,27 +46,132 @@ public function get($params)
         $params['sld'] = $params['original']['domainObj']->getSecondLevel();
         $params['tld'] = $params['original']['domainObj']->getTopLevel();
 
-        $dnsRecordsArr = array();
         $this->domain->load(array(
             'name'      => $params['sld'],
             'extension' => $params['tld']
         ));
 
-        $dnsInfo = $this->apiHelper->getDns($this->domain);
+        try {
+            $dnsInfo = $this->apiHelper->getDns($this->domain);
+        } catch (\Exception $e) {
+            return [];
+        }
 
         if (empty($dnsInfo)) {
             return [];
         }
 
-        $supportedDnsTypes = APIConfig::$supportedDnsTypes;
-        $domainName        = $params['sld'] . '.' . $params['tld'];
+        $domainName = $params['sld'] . '.' . $params['tld'];
+
+        $dnsRecords = $this->getDisplayedRecords($dnsInfo['records']);
+
+        return $this->prepareRecordsToDisplay($dnsRecords, $domainName);
+    }
+
+    /**
+     * Save the new DNS settings.
+     *
+     * @param $params
+     *
+     * @return array|string
+     */
+    public function save($params)
+    {
+        $params['sld'] = $params['original']['domainObj']->getSecondLevel();
+        $params['tld'] = $params['original']['domainObj']->getTopLevel();
+
+        $dnsRecordsArr = $this->prepareRecordsToSave($params['dnsrecords']);
+
+        $domain = $this->domain;
+        $domain->name = $params['sld'];
+        $domain->extension = $params['tld'];
+
+        $values = [];
+        try {
```

Comment:
function processCallback(callable $callback) {
try {
  $callback();
} catch (\Exception $e) {
   return ["error" => $e->getMessage()];
}

return "success";
}

if (empty($dnsRecordsArr)) {
  return updateRecords(function() {
     $this->apiHelper->deleteDnsRecords($domain);
  });
}

$dnsZone = $this->apiHelper->getDns($domain);
if ($dnsZone) {
    return updateRecords(function() {
      $this->apiHelper->updateDnsRecords($domain, $dnsRecordsArr);
  });
} 

return updateRecords(function() {
    $this->apiHelper->createDnsRecords($domain, $dnsRecordsArr);
});
}